### PR TITLE
New Way of Fuzzing Packet

### PR DESCRIPTION
**Pull request by gaurav2699**
## Description

This PR focuses on the the work involving fuzzing of the packets by modifying different bytes of the packet array instead of header values.
## Testing

_Do any existing tests cover this change? Are new tests needed?_

## Documentation

_Is there any documentation impact for this change?_

### CODE REVIEW

File: src/tools/recvfuzz/recvfuzz.cpp [hunk 1]
```diff
@@ -547,17 +563,7 @@ void start() {
...
-
-    auto CredConfig = CxPlatGetSelfSignedCert(CXPLAT_SELF_SIGN_CERT_USER, FALSE, NULL);
-
-    MsQuicConfiguration ServerConfiguration(Registration, Alpn, *CredConfig);
-
-    QUIC_SUCCEEDED(ServerConfiguration.GetInitStatus());
-    MsQuicAutoAcceptListener Listener(Registration, ServerConfiguration, MsQuicConnection::NoOpCallback);
-    QUIC_SUCCEEDED(Listener.Start(Alpn, &sockAddr));
-    QUIC_SUCCEEDED(Listener.GetInitStatus());
+    makeServer(&sockAddr);
```
Comment (nibanks) [inline src/tools/recvfuzz/recvfuzz.cpp:566]: Don't put this in a separate function because when the function returns, everything gets cleaned up that was created in it.

File: src/tools/recvfuzz/recvfuzz.cpp [hunk 2]
```diff
@@ -602,14 +615,14 @@ int
...
-
     uint32_t RngSeed = 0;
     if (!TryGetValue(argc, argv, "seed", &RngSeed)) {
         CxPlatRandom(sizeof(RngSeed), &RngSeed);
     }   
     printf("Using seed value: %u\n", RngSeed);
     srand(RngSeed);
-
+    mode = (uint8_t)GetRandom(2);
+    numIteration = (uint8_t)GetRandom(256);
```
Comment (nibanks) [inline src/tools/recvfuzz/recvfuzz.cpp:625]: These should both be picked per-packet, in the inner loop.

File: src/tools/recvfuzz/recvfuzz.cpp [hunk 3]
```diff
@@ -65,9 +65,7 @@ class FuzzingData {
         return true;
     }
 public:
-    static const size_t UtilityDataSize = 20; // hard code for determinisity
-
-    FuzzingData(const uint8_t* data, size_t size) : data(data), size(size - UtilityDataSize) {}
+    FuzzingData(const uint8_t* data, size_t size) : data(data), size(size) {}
```
Comment (nibanks) [inline src/tools/recvfuzz/recvfuzz.cpp:68]: `size` can be `const` now (up above on line 60).

File: src/tools/recvfuzz/recvfuzz.cpp [hunk 4]
```diff
@@ -87,7 +85,7 @@ static FuzzingData* FuzzData = nullptr;
 template<typename T>
 T GetRandom(T UpperBound) {
     if (!FuzzData) {
-        return (T)(rand() % (int)UpperBound);
+        return (T)(rand() % (int)(UpperBound));
```
Comment (nibanks) [inline src/tools/recvfuzz/recvfuzz.cpp:88]: Unnecessary change.

File: src/tools/recvfuzz/recvfuzz.cpp [hunk 5]
```diff
@@ -398,122 +396,141 @@ void WriteClientInitialPacket(
     *PacketLength += CXPLAT_ENCRYPTION_OVERHEAD;
 }
 
-void fuzzInitialPacket(CXPLAT_SOCKET* Binding, CXPLAT_ROUTE Route) {
+void fuzzPacket(uint8_t* Packet, uint16_t PacketLength) {
+    uint8_t numIteration = (uint8_t)GetRandom(256);
+    for(int i = 0; i < numIteration; i++){
+        uint16_t offset = (uint16_t)GetRandom(PacketLength); // offset
+        uint8_t value = (uint8_t)GetRandom(256); // value
+        *(Packet + offset) = value;
```
Comment (nibanks) [inline src/tools/recvfuzz/recvfuzz.cpp:404]:
```suggestion
        Packet[GetRandom(PacketLength)] = (uint8_t)GetRandom(256);
```

File: src/tools/recvfuzz/recvfuzz.cpp [hunk 6]
```diff
@@ -398,122 +396,141 @@ void WriteClientInitialPacket(
...
-        if (!SendData){
-            printf("CxPlatSendDataAlloc failed\n");
-        }
-        while (CxPlatTimeDiff64(StartTimeMs, CxPlatTimeMs64()) < RunTimeMs && !CxPlatSendDataIsFull(SendData)) {
-            const uint64_t PacketNumber = 0; //fuzz
-            uint8_t Packet[512] = {0};
-            uint16_t PacketLength, HeaderLength;
+    CXPLAT_SEND_CONFIG SendConfig = { &Route, DatagramLength, CXPLAT_ECN_NON_ECT, 0 };
+    CXPLAT_SEND_DATA* SendData = CxPlatSendDataAlloc(Binding, &SendConfig);
+    if (!SendData){
```
Comment (nibanks) [inline src/tools/recvfuzz/recvfuzz.cpp:413]:
```suggestion
    if (!SendData) {
```

File: src/tools/recvfuzz/recvfuzz.cpp [hunk 7]
```diff
@@ -398,122 +396,141 @@ void WriteClientInitialPacket(
...
-            InterlockedExchangeAdd64(&TotalByteCount, DatagramLength);
         }
+        if (!OrigSrcCid) {
+            printf("Failed to find OrigSrcCid!\n");
+            return;
+        }
+
+        CxPlatRandom(sizeof(uint64_t), DestCid); //fuzz
+        CxPlatRandom(sizeof(uint64_t), SrcCid); //fuzz
+        if(fuzzing){
```
Comment (nibanks) [inline src/tools/recvfuzz/recvfuzz.cpp:447]:
```suggestion
        if (fuzzing){
```

File: src/tools/recvfuzz/recvfuzz.cpp [hunk 8]
```diff
@@ -398,122 +396,141 @@ void WriteClientInitialPacket(
...
+    if (QUIC_FAILED(
+        CxPlatSocketSend(
+            Binding,
+            &Route,
+            SendData))) {
+        printf("Send failed!\n");
+        exit(0);
+    }
+
+}
```
Comment (nibanks) [inline src/tools/recvfuzz/recvfuzz.cpp:510]:
```suggestion
    }
}
```

File: src/tools/recvfuzz/recvfuzz.cpp [hunk 9]
```diff
@@ -398,122 +396,141 @@ void WriteClientInitialPacket(
...
+            fuzzInitialPacket(Binding, Route, StartTimeMs, &PacketCount, &TotalByteCount);
+        } else if (mode == 1){
+            printf("Running Handshake Packet Fuzzing\n");
+            fuzzHandshakePacket();
+        }
+    }
+        printf("Total Packets sent: %lld\n", (long long)PacketCount);
+        printf("Total Bytes sent: %lld\n", (long long)TotalByteCount);
+
 }
```
Comment (nibanks) [inline src/tools/recvfuzz/recvfuzz.cpp:534]:
```suggestion
    printf("Total Packets sent: %lld\n", (long long)PacketCount);
    printf("Total Bytes sent: %lld\n", (long long)TotalByteCount);
}
```

File: src/tools/recvfuzz/recvfuzz.cpp [hunk 10]
```diff
@@ -398,122 +396,141 @@ void WriteClientInitialPacket(
...
+    int64_t PacketCount = 0;
+    int64_t TotalByteCount = 0;
+    uint8_t mode;
+    uint64_t StartTimeMs = CxPlatTimeMs64();
+    while (CxPlatTimeDiff64(StartTimeMs, CxPlatTimeMs64()) < RunTimeMs) {
+        mode = (uint8_t)GetRandom(2);
+        if (mode == 0) {
+            printf("Running Initial Packet Fuzzing\n");
+            fuzzInitialPacket(Binding, Route, StartTimeMs, &PacketCount, &TotalByteCount);
+        } else if (mode == 1){
```
Comment (nibanks) [inline src/tools/recvfuzz/recvfuzz.cpp:526]:
```suggestion
        } else if (mode == 1) {
```

File: src/tools/recvfuzz/recvfuzz.cpp [hunk 11]
```diff
@@ -398,122 +396,141 @@ void WriteClientInitialPacket(
...
+// TODO
+}
+
+void fuzz(CXPLAT_SOCKET* Binding, CXPLAT_ROUTE Route) {
+    int64_t PacketCount = 0;
+    int64_t TotalByteCount = 0;
+    uint8_t mode;
+    uint64_t StartTimeMs = CxPlatTimeMs64();
+    while (CxPlatTimeDiff64(StartTimeMs, CxPlatTimeMs64()) < RunTimeMs) {
+        mode = (uint8_t)GetRandom(2);
```
Comment (nibanks) [inline src/tools/recvfuzz/recvfuzz.cpp:522]: Until `mode == 1` is supported (in a future PR) let's always do `mode == 0`.

File: src/tools/recvfuzz/recvfuzz.cpp [hunk 12]
```diff
@@ -398,122 +396,141 @@ void WriteClientInitialPacket(
...
+        if(fuzzing){
+            fuzzPacket(Packet, sizeof(Packet));
+        }
+        QUIC_BUFFER* SendBuffer =
+            CxPlatSendDataAllocBuffer(SendData, DatagramLength);
+            if (!SendBuffer) {
+            printf("CxPlatSendDataAllocBuffer failed\n");
+            return;
+            }
+        (*DestCid)++; (*SrcCid)++;
```
[the remark this answers is not in the document]
Comment (gaurav2699) [reply src/tools/recvfuzz/recvfuzz.cpp:456]: Yep, dont think its necessary

**Comment by nibanks** [comment]
Your tool hit this:
```
==4188==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x00f7a58fef87 at pc 0x7ff67d2aac5f bp 0x00f7a58feb30 sp 0x00f7a58feb30
READ of size 1 at 0x00f7a58fef87 thread T0
    #0 0x7ff67d2aac5e in QuicCryptoCombineIvAndPacketNumber D:\a\msquic\msquic\src\inc\quic_crypt.h:275
    #1 0x7ff67d2a2676 in buildInitialPacket(struct CXPLAT_SOCKET *, struct CXPLAT_ROUTE, __int64 *, __int64 *, bool) D:\a\msquic\msquic\src\tools\recvfuzz\recvfuzz.cpp:471
    #2 0x7ff67d2a2ce1 in fuzz(struct CXPLAT_SOCKET *, struct CXPLAT_ROUTE) D:\a\msquic\msquic\src\tools\recvfuzz\recvfuzz.cpp:520
    #3 0x7ff67d2a3822 in start(void) D:\a\msquic\msquic\src\tools\recvfuzz\recvfuzz.cpp:598
    #4 0x7ff67d2a3a[13](https://github.com/microsoft/msquic/actions/runs/6695229133/job/18190860302?pr=3942#step:7:14) in main D:\a\msquic\msquic\src\tools\recvfuzz\recvfuzz.cpp:622
    #5 0x7ff67d4541fb in invoke_main D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78
    #6 0x7ff67d4541fb in __scrt_common_main_seh D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
    #7 0x7ffac7244ddf  (C:\Windows\System32\KERNEL32.DLL+0x1800[14](https://github.com/microsoft/msquic/actions/runs/6695229133/job/18190860302?pr=3942#step:7:15)ddf)
    #8 0x7ffac7cfed9a  (C:\Windows\SYSTEM32\ntdll.dll+0x[18](https://github.com/microsoft/msquic/actions/runs/6695229133/job/18190860302?pr=3942#step:7:19)007ed9a)
```

File: src/tools/recvfuzz/recvfuzz.cpp [hunk 13]
```diff
@@ -398,122 +396,134 @@ void WriteClientInitialPacket(
...
-            const uint64_t PacketNumber = 0; //fuzz
-            uint8_t Packet[512] = {0};
-            uint16_t PacketLength, HeaderLength;
+    CXPLAT_SEND_CONFIG SendConfig = { &Route, DatagramLength, CXPLAT_ECN_NON_ECT, 0 };
+    CXPLAT_SEND_DATA* SendData = CxPlatSendDataAlloc(Binding, &SendConfig);
+    if (!SendData) {
+        printf("CxPlatSendDataAlloc failed\n");
+    }
+    while (!CxPlatSendDataIsFull(SendData)) {
+        const uint32_t PacketNumber = GetRandom(1000);
```
Comment (nibanks) [inline src/tools/recvfuzz/recvfuzz.cpp:415]:
```suggestion
        const uint64_t PacketNumber = GetRandom(1000);
```

Comment (gaurav2699) [reply src/tools/recvfuzz/recvfuzz.cpp]: I believe your recent commit fixes this

**Review by nibanks — APPROVED** [review]
